File: kryos/lsp/protocol.py
"""
Kryos LSP -- JSON-RPC message parsing/encoding and LSP type definitions.

Implements the subset of the Language Server Protocol needed by the Kryos
language server.  Pure Python stdlib -- no pip dependencies.
"""

from __future__ import annotations

import json
import sys
from dataclasses import dataclass, field, asdict
from typing import Any, Dict, List, Optional, Union


# ---------------------------------------------------------------------------
# JSON-RPC transport
# ---------------------------------------------------------------------------

def read_message(stream=None) -> Optional[Dict[str, Any]]:
    """
    Read a single JSON-RPC message from *stream* (default ``sys.stdin.buffer``).

    Returns the parsed JSON dict, or ``None`` on EOF.
    """
    if stream is None:
        stream = sys.stdin.buffer

    # Read headers until blank line
    headers: Dict[str, str] = {}
    while True:
        line = stream.readline()
        if not line:
            return None  # EOF
        line_str = line.decode("ascii", errors="replace").rstrip("\r\n")
        if line_str == "":
            break  # end of headers
        if ":" in line_str:
            key, _, val = line_str.partition(":")
            headers[key.strip().lower()] = val.strip()

    content_length = int(headers.get("content-length", "0"))
    if content_length == 0:
        return None

    body = stream.read(content_length)
    if not body:
        return None

    return json.loads(body.decode("utf-8"))


def write_message(msg: Dict[str, Any], stream=None) -> None:
    """
    Write a single JSON-RPC message to *stream* (default ``sys.stdout.buffer``).
    """
    if stream is None:
        stream = sys.stdout.buffer

    body = json.dumps(msg, ensure_ascii=False).encode("utf-8")
    header = f"Content-Length: {len(body)}\r\n\r\n".encode("ascii")
    stream.write(header)
    stream.write(body)
    stream.flush()


def make_response(req_id: Union[int, str, None], result: Any) -> Dict[str, Any]:
    """Build a JSON-RPC success response."""
    return {"jsonrpc": "2.0", "id": req_id, "result": result}


def make_error(req_id: Union[int, str, None], code: int, message: str,
               data: Any = None) -> Dict[str, Any]:
    """Build a JSON-RPC error response."""
    err: Dict[str, Any] = {"code": code, "message": message}
    if data is not None:
        err["data"] = data
    return {"jsonrpc": "2.0", "id": req_id, "error": err}


def make_notification(method: str, params: Any) -> Dict[str, Any]:
    """Build a JSON-RPC notification (no id)."""
    return {"jsonrpc": "2.0", "method": method, "params": params}


# ---------------------------------------------------------------------------
# JSON-RPC error codes
# ---------------------------------------------------------------------------

PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INVALID_PARAMS = -32602
INTERNAL_ERROR = -32603
SERVER_NOT_INITIALIZED = -32002
REQUEST_CANCELLED = -32800


# ---------------------------------------------------------------------------
# LSP Position / Range / Location
# ---------------------------------------------------------------------------

@dataclass
class Position:
    """Zero-based line and character offset."""
    line: int = 0
    character: int = 0

    def to_dict(self) -> Dict[str, int]:
        return {"line": self.line, "character": self.character}


@dataclass
class Range:
    """A range in a text document (start inclusive, end exclusive)."""
    start: Position = field(default_factory=Position)
    end: Position = field(default_factory=Position)

    def to_dict(self) -> Dict[str, Any]:
        return {"start": self.start.to_dict(), "end": self.end.to_dict()}


@dataclass
class Location:
    uri: str = ""
    range: Range = field(default_factory=Range)

    def to_dict(self) -> Dict[str, Any]:
        return {"uri": self.uri, "range": self.range.to_dict()}


# ---------------------------------------------------------------------------
# Diagnostics
# ---------------------------------------------------------------------------

class DiagnosticSeverity:
    Error = 1
    Warning = 2
    Information = 3
    Hint = 4


@dataclass
class Diagnostic:
    range: Range = field(default_factory=Range)
    severity: int = DiagnosticSeverity.Error
    source: str = "kryos"
    message: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "range": self.range.to_dict(),
            "severity": self.severity,
            "source": self.source,
            "message": self.message,
        }


# ---------------------------------------------------------------------------
# Completion
# ---------------------------------------------------------------------------

class CompletionItemKind:
    Text = 1
    Method = 2
    Function = 3
    Constructor = 4
    Field = 5
    Variable = 6
    Class = 7
    Interface = 8
    Module = 9
    Property = 10
    Unit = 11
    Value = 12
    Enum = 13
    Keyword = 14
    Snippet = 15
    Color = 16
    File = 17
    Reference = 18
    Folder = 19
    EnumMember = 20
    Constant = 21
    Struct = 22
    Event = 23
    Operator = 24
    TypeParameter = 25


@dataclass
class CompletionItem:
    label: str = ""
    kind: int = CompletionItemKind.Text
    detail: str = ""
    documentation: str = ""
    insert_text: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "label": self.label,
            "kind": self.kind,
        }
        if self.detail:
            d["detail"] = self.detail
        if self.documentation:
            d["documentation"] = {"kind": "markdown", "value": self.documentation}
        if self.insert_text is not None:
            d["insertText"] = self.insert_text
        return d


# ---------------------------------------------------------------------------
# Document Symbols
# ---------------------------------------------------------------------------

class SymbolKind:
    File = 1
    Module = 2
    Namespace = 3
    Package = 4
    Class = 5
    Method = 6
    Property = 7
    Field = 8
    Constructor = 9
    Enum = 10
    Interface = 11
    Function = 12
    Variable = 13
    Constant = 14
    String = 15
    Number = 16
    Boolean = 17
    Array = 18
    Object = 19
    Key = 20
    Null = 21
    EnumMember = 22
    Struct = 23
    Event = 24
    Operator = 25
    TypeParameter = 26


@dataclass
class DocumentSymbol:
    name: str = ""
    kind: int = SymbolKind.Function
    range: Range = field(default_factory=Range)
    selection_range: Range = field(default_factory=Range)
    detail: str = ""
    children: List["DocumentSymbol"] = field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "name": self.name,
            "kind": self.kind,
            "range": self.range.to_dict(),
            "selectionRange": self.selection_range.to_dict(),
        }
        if self.detail:
            d["detail"] = self.detail
        if self.children:
            d["children"] = [c.to_dict() for c in self.children]
        return d


# ---------------------------------------------------------------------------
# Hover
# ---------------------------------------------------------------------------

@dataclass
class Hover:
    contents: str = ""
    range: Optional[Range] = None

    def to_dict(self) -> Dict[str, Any]:
        d: Dict[str, Any] = {
            "contents": {"kind": "markdown", "value": self.contents},
        }
        if self.range is not None:
            d["range"] = self.range.to_dict()
        return d


# ---------------------------------------------------------------------------
# TextEdit helpers
# ---------------------------------------------------------------------------

class TextDocumentSyncKind:
    None_ = 0
    Full = 1
    Incremental = 2

File: compiler/crates/kryos-lsp/src/protocol.rs
//! JSON-RPC protocol types for LSP communication.

use serde_json::Value;

/// JSON-RPC request.
#[derive(Debug, Clone)]
pub struct Request {
    pub id: Value,
    pub method: String,
    pub params: Value,
}

/// JSON-RPC response.
#[derive(Debug, Clone)]
pub struct Response {
    pub id: Value,
    pub result: Option<Value>,
    pub error: Option<ResponseError>,
}

#[derive(Debug, Clone)]
pub struct ResponseError {
    pub code: i32,
    pub message: String,
    pub data: Option<Value>,
}

/// JSON-RPC notification (no id, no response expected).
#[derive(Debug, Clone)]
pub struct Notification {
    pub method: String,
    pub params: Value,
}

/// LSP message types.
#[derive(Debug)]
pub enum Message {
    Request(Request),
    Response(Response),
    Notification(Notification),
}

/// Parse a JSON-RPC message from a JSON value.
pub fn parse_message(value: Value) -> Option<Message> {
    let obj = value.as_object()?;
    let method = obj.get("method").and_then(|m| m.as_str()).map(String::from);
    let id = obj.get("id").cloned();
    let params = obj.get("params").cloned().unwrap_or(Value::Null);

    match (id, method) {
        (Some(id), Some(method)) => Some(Message::Request(Request { id, method, params })),
        (None, Some(method)) => Some(Message::Notification(Notification { method, params })),
        (Some(id), None) => {
            let result = obj.get("result").cloned();
            let error = obj.get("error").and_then(|e| {
                let eo = e.as_object()?;
                Some(ResponseError {
                    code: eo.get("code")?.as_i64()? as i32,
                    message: eo.get("message")?.as_str()?.to_string(),
                    data: eo.get("data").cloned(),
                })
            });
            Some(Message::Response(Response { id, result, error }))
        }
        _ => None,
    }
}

/// Serialize a response to JSON.
pub fn make_response(id: Value, result: Value) -> Value {
    serde_json::json!({
        "jsonrpc": "2.0",
        "id": id,
        "result": result,
    })
}

/// Serialize an error response to JSON.
pub fn make_error_response(id: Value, code: i32, message: &str) -> Value {
    serde_json::json!({
        "jsonrpc": "2.0",
        "id": id,
        "error": {
            "code": code,
            "message": message,
        },
    })
}

/// Serialize a notification to JSON.
pub fn make_notification(method: &str, params: Value) -> Value {
    serde_json::json!({
        "jsonrpc": "2.0",
        "method": method,
        "params": params,
    })
}

/// Read a single LSP message from a reader (Content-Length header + JSON body).
pub fn read_message(reader: &mut impl std::io::BufRead) -> std::io::Result<Option<Value>> {
    let mut content_length: Option<usize> = None;
    let mut header_line = String::new();

    // Read headers
    loop {
        header_line.clear();
        let bytes_read = reader.read_line(&mut header_line)?;
        if bytes_read == 0 {
            return Ok(None); // EOF
        }
        let line = header_line.trim();
        if line.is_empty() {
            break; // End of headers
        }
        if let Some(len_str) = line.strip_prefix("Content-Length: ") {
            content_length = len_str.parse().ok();
        }
    }

    let length = content_length.ok_or_else(|| {
        std::io::Error::new(std::io::ErrorKind::InvalidData, "missing Content-Length")
    })?;

    let mut body = vec![0u8; length];
    reader.read_exact(&mut body)?;

    let value: Value = serde_json::from_slice(&body)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;

    Ok(Some(value))
}

/// Write an LSP message to a writer.
pub fn write_message(writer: &mut impl std::io::Write, value: &Value) -> std::io::Result<()> {
    let body = serde_json::to_string(value)?;
    write!(writer, "Content-Length: {}\r\n\r\n{}", body.len(), body)?;
    writer.flush()
}
